### PR TITLE
Users/prmutn/specifylocationnode

### PR DESCRIPTION
**Pull request by preetsaimutneja**
Allow users to specify vstest.console.exe path.

1. Pending unit tests
2. Did manual testing

### CODE REVIEW

File: Tasks/VsTest/vstest.ts [hunk 1]
```diff
@@ -245,13 +247,25 @@ function publishCodeChanges(): Q.Promise<string> {
 
 function executeVstest(testResultsDirectory: string, parallelRunSettingsFile: string, vsVersion: number, argsArray: string[]): Q.Promise<number> {
     var defer = Q.defer<number>();
-    var vsCommon = tl.getVariable("VS" + vsVersion + "0COMNTools");
-    if (!vsCommon) {
-        tl.error(tl.loc('VstestNotFound', vsVersion));
-        defer.resolve(1);
-        return defer.promise;
+    if (vstestLocationMethod == "version") {
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: === instead of ==
always use toLowercase and compare

Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: split into multiple functions. not more than 10 lines for function

File: Tasks/VsTest/vstest.ts [hunk 2]
```diff
@@ -245,13 +247,25 @@ function publishCodeChanges(): Q.Promise<string> {
...
+        var vsCommon = tl.getVariable("VS" + vsVersion + "0COMNTools");
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: always use `let`

File: Tasks/VsTest/vstest.ts [hunk 3]
```diff
@@ -245,13 +247,25 @@ function publishCodeChanges(): Q.Promise<string> {
...
+        if (!vsCommon) {
+            tl.error(tl.loc('VstestNotFound', vsVersion));
+            defer.resolve(1);
+            return defer.promise;
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: avoid defer. return directly `Q.resolve`

File: Tasks/VsTest/vstest.ts [hunk 4]
```diff
@@ -245,13 +247,25 @@ function publishCodeChanges(): Q.Promise<string> {
...
+        }
+        vstestLocation = path.join(vsCommon, "..\\IDE\\CommonExtensions\\Microsoft\\TestWindow\\vstest.console.exe");
+    } else if (vstestLocationMethod == "location") {
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: single quotes and ===

File: Tasks/VsTest/vstest.ts [hunk 5]
```diff
@@ -245,13 +247,25 @@ function publishCodeChanges(): Q.Promise<string> {
...
+        try {
+            fs.accessSync(vstestLocation);
+        } catch (error) {
+            tl.error(tl.loc('AccessDeniedToPath', vstestLocation));
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: same here

File: Tasks/VsTest/vstest.ts [hunk 6]
```diff
@@ -245,13 +247,25 @@ function publishCodeChanges(): Q.Promise<string> {
...
+            defer.resolve(1);
+            return defer.promise;
+        }
+    } else {
+        ;
```
Comment (vimegh) [inline Tasks/VsTest/vstest.ts]: Empty else?

**Comment by vimegh** [comment]
Patch no. needs to be incremented by 1

File: Tasks/VsTest/vstest.ts [hunk 7]
```diff
@@ -245,13 +247,25 @@ function publishCodeChanges(): Q.Promise<string> {
...
+        var vsCommon = tl.getVariable("VS" + vsVersion + "0COMNTools");
+        if (!vsCommon) {
+            tl.error(tl.loc('VstestNotFound', vsVersion));
+            defer.resolve(1);
+            return defer.promise;
+        }
+        vstestLocation = path.join(vsCommon, "..\\IDE\\CommonExtensions\\Microsoft\\TestWindow\\vstest.console.exe");
+    } else if (vstestLocationMethod == "location") {
+        try {
+            fs.accessSync(vstestLocation);
```
Comment (vimegh) [inline Tasks/VsTest/vstest.ts]: Will fs.accessSync throw error if File Doesn't exist? if yes then file not there and access denied will lead to same exception

Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: actually. you should use `tl.exists()` there is already lib function.

File: Tasks/VsTest/vstest.ts [hunk 8]
```diff
@@ -245,13 +247,21 @@ function publishCodeChanges(): Q.Promise<string> {
 
 function executeVstest(testResultsDirectory: string, parallelRunSettingsFile: string, vsVersion: number, argsArray: string[]): Q.Promise<number> {
     var defer = Q.defer<number>();
-    var vsCommon = tl.getVariable("VS" + vsVersion + "0COMNTools");
-    if (!vsCommon) {
-        tl.error(tl.loc('VstestNotFound', vsVersion));
-        defer.resolve(1);
-        return defer.promise;
+    if (vstestLocationMethod.toLowerCase() === 'version') {
+        let vsCommon = tl.getVariable("VS" + vsVersion + "0COMNTools");
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: this should be separate method

**Comment by nigurr** [comment]
rest looks good

**Comment by vimegh** [comment]
:shipit:

**Comment by vimegh** [comment]
merge the conflicts.

**Pull request merged**